modal/_container_io_manager.py

```diff
@@ -401,48 +420,142 @@ async def _generate_inputs(self) -> AsyncIterator[Tuple[str, str, api_pb2.Functi
                         if item.input.final_input or self.function_def.max_inputs == 1:
                             eof_received = True
                             break
+                    else:
+                        assert len(response.inputs) <= request.batch_max_size
+
+                        inputs_list = []
+                        for item in response.inputs:
+                            if item.kill_switch:
+                                assert len(response.inputs) == 1
+                                logger.debug(f"Task {self.task_id} input kill signal input.")
+                                eof_received = True
+                                break
+                            assert item.input_id not in self.cancelled_input_ids
+
+                            # If we got a pointer to a blob, download it from S3.
+                            if item.input.WhichOneof("args_oneof") == "args_blob_id":
+                                input_pb = await self.populate_input_blobs(item.input)
+                            else:
+                                input_pb = item.input
+
+                            inputs_list.append((item.input_id, item.function_call_id, input_pb))
+                            if item.input.final_input:
+                                eof_received = True
+                                logger.error("Final input not expected in batched input stream")
+                                break
+
+                        if not eof_received:
+                            yield inputs_list
+                            yielded = True
+
+                            # We only support max_inputs = 1 at the moment
+                            if self.function_def.max_inputs == 1:
+                                eof_received = True
+
             finally:
                 if not yielded:
                     self._semaphore.release()
 
     @synchronizer.no_io_translation
-    async def run_inputs_outputs(self, input_concurrency: int = 1) -> AsyncIterator[LocalInput]:
+    async def run_inputs_outputs(
```

Comment:
Just a thought but would it be cleaner to have separate functions for batched / nonbatched?